pkg/addon/utils.go

```diff
@@ -0,0 +1,201 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package addon
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	errors "github.com/pkg/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	rest "k8s.io/client-go/rest"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/pkg/definition"
+	"github.com/oam-dev/kubevela/pkg/oam"
+	"github.com/oam-dev/kubevela/pkg/oam/util"
+)
+
+const (
+	compDefAnnotation         = "addon.oam.dev/componentDefinitions"
+	traitDefAnnotation        = "addon.oam.dev/traitDefinitions"
+	workflowStepDefAnnotation = "addon.oam.dev/workflowStepDefinitions"
+	defKeytemplate            = "addon-%s-%s"
+)
+
+// parse addon's created x-defs in addon-app's annotation, this will be used to check whether app still using it while disabling.
+func passDefInAppAnnotation(defs []*unstructured.Unstructured, app *v1beta1.Application) error {
+	var comps, traits, workflowSteps []string
+	for _, def := range defs {
+		switch def.GetObjectKind().GroupVersionKind().Kind {
+		case v1beta1.ComponentDefinitionKind:
+			comps = append(comps, def.GetName())
+		case v1beta1.TraitDefinitionKind:
+			traits = append(traits, def.GetName())
+		case v1beta1.WorkflowStepDefinitionKind:
+			workflowSteps = append(workflowSteps, def.GetName())
+		default:
+			return fmt.Errorf("cannot handle definition types %s, name %s", def.GetObjectKind().GroupVersionKind().Kind, def.GetName())
+		}
+	}
+	if len(comps) != 0 {
+		c, _ := json.Marshal(comps)
```

Comment:
we should narrow the length, I suggest just use comma to sperate:

`traitAnnotation: mytrait1,mytrait2,mytrait3`

Reply:
good suggestion. fixed